### ISSUE TITLE
If allocate-ipip-addr fails (due to expected race), it doesn't get retried

### ISSUE THREAD

**Issue by ti-mo** [projectcalico/node#9]
Mentioned in https://github.com/projectcalico/libcalico-go/pull/785.

## Expected Behavior

Declaring an IP pool on an existing deployment should allocate a block to every cluster member.

## Current Behavior

```
2018-04-27T08:49:12Z public-slave-14 confd[130]: ERROR "
2018-04-27 08:49:11.878 [INFO][277] client.go 202: Loading config from environment
2018-04-27 08:49:11.895 [INFO][277] ipam.go 120: Auto-assign 1 ipv4, 0 ipv6 addrs for host 'public-slave-14'
2018-04-27 08:49:11.903 [INFO][277] ipam.go 172: Ran out of existing affine blocks for host 'public-slave-14'
2018-04-27 08:49:11.909 [INFO][277] ipam.go 195: Need to allocate 1 more addresses - allocate another block
2018-04-27 08:49:11.932 [INFO][277] ipam_block_reader_writer.go 116: Claiming a new affine block for host 'public-slave-14'
2018-04-27 08:49:11.944 [INFO][277] ipam_block_reader_writer.go 159: Host public-slave-14 claiming block affinity for 10.123.84.64/26
2018-04-27 08:49:11.994 [WARNING][277] ipam_block_reader_writer.go 187: Problem claiming block affinity:%!!(MISSING)(EXTRA errors.ErrorResourceAlreadyExists=resource already exists: BlockKey(cidr=10.123.84.64/26))
2018-04-27 08:49:12.060 [ERROR][277] ipam.go 204: Error claiming new block: 10.123.84.64/26 already claimed by %!!(MISSING)s(*string=0xc420283430)
2018-04-27 08:49:12.060 [ERROR][277] ipam.go 135: Error assigning IPV4 addresses: 10.123.84.64/26 already claimed by %!!(MISSING)s(*string=0xc420283430)
2018-04-27 08:49:12.060 [FATAL][277] allocate_ipip_addr.go 120: Unable to autoassign an address for IPIP error=10.123.84.64/26 already claimed by %!!(MISSING)s(*string=0xc420283430)
"
2018-04-27T08:49:12Z public-slave-14 confd[130]: ERROR exit status 1
```

Unlike we initially assumed, `runsv` _does_ successfully restart confd when it crashes (obviously), but confd does not re-render a template when its `reload_cmd` fails. (and I don't think it should, either) The `ERROR exit status 1` is from the `allocate-ipip-addr` subprocess run by confd.

## Possible Solution

Of course this is completely up to you, but to me it looks like a retry mechanism would be the most straightforward to implement. Locking is already in place, since one block clearly cannot be claimed by multiple hosts.

Could be done on two levels, either
- around the `clientv3.IPAM().AutoAssign()` at https://github.com/projectcalico/node/blob/master/pkg/allocateipip/allocate_ipip_addr.go#L162, or
- inside `AutoAssign()` itself

## Steps to Reproduce (for bugs)

1. Spin up a couple dozen `calico/node` instances connected to etcd
2. Declare an ipPool (**with ipip!**)
3. Check if every node has an address assigned to `tunl0` (this might go undetected in tests); alternatively, check `/var/log/confd/current` for error output
4. Rinse and repeat if necessary, this happens about 80% of the time on at least one node in our larger clusters

Note: I assume this could be tested in isolation by spawning a hundred `allocate-ipip-addr` simultaneously on a single box. This is not super cheap (money-wise) to replicate on a fleet of VMs.

## Your Environment

Some details I did not mention in the original post:
- Cluster size: about 40 nodes
- Pool size: `/16`
- Block size: `/26` (default?)
- `calico/node` 2.6.7
- etcd back-end

The ipPool has ipip enabled:
```
spec:
  ipip:
    enabled: true
    mode: always
  nat-outgoing: true
```

@caseydavenport @fasaxc 

Thanks in advance for your time and for developing Calico!

**Comment by ti-mo** [comment]
@caseydavenport Wouldn't the work you did in https://github.com/projectcalico/libcalico-go/commit/28b80c767d49a50d5a990e9a58a1a5c5c406e322 solve this? I realize this might be incompatible with 2.6 (on top of introducing a large change into a maintenance branch), so should we just go for a simple retry in `allocate-ipip-addr` itself? We would upgrade to 3.0 if we could, but yeah.. Mesos.. :smile:

**Comment by caseydavenport** [comment]
@ti-mo yes, I believe that work _would_ solve this.

I think it's too hard a back-port that fix onto v2.6, so I'm happy with a patch to add a retry to `allocate-ipip-addr`

**Comment by caseydavenport** [comment]
Fixed in https://github.com/projectcalico/calico/pull/2038

Should be in v2.6.11 when it comes out.  Thanks @ti-mo!
